src/pkcs11.c

```diff
@@ -900,7 +900,9 @@ int pkcs11_module_load(struct pkcs11_module *pkcs,
 
 close_err:
 	dlclose(pkcs->dlhandle);
+	pkcs->dlhandle = NULL;
```

Comment:
Alternatively you could just omit any `dlclose`/`OPENSSL_free` herein, since on failure the caller calls `ps_prov_teardown()` which in turn calls `pkcs11_module_teardown()` which frees/dlcloses anything that is left over anyway. 

You are not freeing `pkcs->initargs` in the error path of `pkcs11_module_load()` anyway, so this is left to be freed for `pkcs11_module_teardown()` in any case.

While at it, there is no error checking for the 2 `OPENSSL_strdup()` calls in `pkcs11_module_load()`. What happens if they return NULL ?

Reply:
You could just call pkcs11_module_teardown() in the error path.

Reply:
I'll remove all cleanups in `pkcs11_module_load()` and add check for the both `OPENSSL_strdup()` calls.